### PR TITLE
[FIX-3900][server] kill multi yarn app in one job

### LINKED ISSUE #3900: [Bug] Kill YarnJob may be fail in cluster？
Forgive my poor english
**Describe the question**
A clear and concise description of what the question is.
When kill yarn task, dolphinscheduler may shows the kill is successful, but the tasks may still running on the yarn cluster
* * *
当 kill yarn 类型的任务的时候，dolphinscheduler 上显示任务状态 kill 成功，但实际任务可能还在 yarn 集群上运行，没有 kill 掉

**Which version of DolphinScheduler:**
 -[1.3.3-release]

**Additional context**
code in org.apache.dolphinscheduler.server.utils.ProcessUtils.cancelApplication 
image
String appid = appIds.get(appIds.size() - 1); 
get the last appId，If the task has multiple stages，and hive.exec.parallel = true，stages will be executed in parallel if there is no dependency between them. If the last job for stage is successful(last appId), but other is running, the task will kill fail in yarn cluster, but dolphinscheduler show success.
* * *
上面的代码是获取 appId 列表中的最后一个进行 kill，如果这个任务有多个stage，并且stage之间没有依赖啊，hive.exec.parallel = true 设置可以并行执行，就会存在最后一个 appId 对应的 job 执行成功，但是中间或者还存在没有执行完的，如果这时候在 dolphinscheduler 将任务 kill 掉，那么 kill 最后一个 appId 没法在集群上 kill 掉任务的，因为最后一个 appId 对应的 job 已经成功，但是还存在其它并行执行的任务

**Requirement or improvement**
- Please describe about your requirements or improvement suggestions.

I haven't tried it on dolphinscheduler，when study dolphinscheduler code, this killYarnJob code is similar to our self-developed scheduler，we meet this question in our system
* * *
我没有在 dolphinscheduler 上试过，但是我在看相关代码的时候，发现和我们自研的调度系统 killYarnJob 的方式是差不多一样得，都是获取最后一个 appId 进行 kill，就会产生这个问题

### PR DESCRIPTION
**Pull request by Eights-Li**
## What is the purpose of the pull request

*fix #3900 kill multi yarn app in one job*

## Brief change log

  - *modify server/utils/ProcessUtils.java  **cancelApplication** method*

## Verify this pull request

  - *server/utils/ProcessUtilsTest to verify this pr*

### CODE REVIEW

File: dolphinscheduler-server/src/main/java/org/apache/dolphinscheduler/server/utils/ProcessUtils.java [hunk 1]
```diff
@@ -38,368 +41,385 @@
...
+     * kill yarn application.
+     *
+     * @param appIds      app id list
+     * @param logger      logger
+     * @param tenantCode  tenant code
+     * @param executePath execute path
+     */
+    public static void cancelApplication(List<String> appIds, Logger logger, String tenantCode, String executePath) {
+        if (CollectionUtils.isNotEmpty(appIds)) {
+
```
Comment (lenboo) [inline dolphinscheduler-server/src/main/java/org/apache/dolphinscheduler/server/utils/ProcessUtils.java:286]: i think we should reduce its cognitive complexity here.

**Review by lenboo — COMMENTED** [review]
some code smell can be wiped out.

File: dolphinscheduler-server/src/main/java/org/apache/dolphinscheduler/server/utils/ProcessUtils.java [hunk 2]
```diff
@@ -38,368 +41,406 @@
...
+     * check is allow ambiguous commands
+     *
+     * @param security security manager
+     * @return allow ambiguous command flag
+     */
+    private static boolean isAllowAmbiguousCommands(SecurityManager security) {
+        boolean allowAmbiguousCommands = false;
+        if (security == null) {
+            allowAmbiguousCommands = true;
+            String value = System.getProperty("jdk.lang.Process.allowAmbiguousCommands");
```
Comment (lenboo) [inline dolphinscheduler-server/src/main/java/org/apache/dolphinscheduler/server/utils/ProcessUtils.java:118]: magic number here?

File: dolphinscheduler-server/src/main/java/org/apache/dolphinscheduler/server/utils/ProcessUtils.java [hunk 3]
```diff
@@ -38,368 +41,406 @@
...
-        cmd = getTokensFromCommand(join.toString());
-        executablePath = getExecutablePath(cmd[0]);
+    /**
+     * get executable path.
+     *
+     * @param path path
+     * @return executable path
+     */
+    private static String getExecutablePath(String path) {
+        boolean pathIsQuoted = isQuoted(true, path, "Executable name has embedded quote, split the arguments");
```
Comment (lenboo) [inline dolphinscheduler-server/src/main/java/org/apache/dolphinscheduler/server/utils/ProcessUtils.java:133]: also magic number.

**Review by lenboo — COMMENTED** [review]
two magic number.

File: dolphinscheduler-server/src/main/java/org/apache/dolphinscheduler/server/utils/ProcessUtils.java (hunk 3 above)
Comment (Eights-Li) [reply dolphinscheduler-server/src/main/java/org/apache/dolphinscheduler/server/utils/ProcessUtils.java:133]: This string is error message

File: dolphinscheduler-server/src/main/java/org/apache/dolphinscheduler/server/utils/ProcessUtils.java (hunk 2 above)
Comment (Eights-Li) [reply dolphinscheduler-server/src/main/java/org/apache/dolphinscheduler/server/utils/ProcessUtils.java]: fixed

File: dolphinscheduler-server/src/main/java/org/apache/dolphinscheduler/server/utils/ProcessUtils.java (hunk 3 above)
Comment (Eights-Li) [reply dolphinscheduler-server/src/main/java/org/apache/dolphinscheduler/server/utils/ProcessUtils.java:135]: fixed

File: dolphinscheduler-server/src/main/java/org/apache/dolphinscheduler/server/utils/ProcessUtils.java [hunk 4]
```diff
@@ -38,368 +41,408 @@
...
+            sb.append("\n\n");
+            sb.append(cmd);
+
+            File f = new File(commandFile);
+
+            if (!f.exists()) {
+                FileUtils.writeStringToFile(new File(commandFile), sb.toString(), StandardCharsets.UTF_8);
+            }
+
+            String runCmd = "sh " + commandFile;
```
Comment (lenboo) [inline dolphinscheduler-server/src/main/java/org/apache/dolphinscheduler/server/utils/ProcessUtils.java:342]: is this "sh " can be instead?
I can find it elsewhere.

File: dolphinscheduler-server/src/main/java/org/apache/dolphinscheduler/server/utils/ProcessUtils.java [hunk 5]
```diff
@@ -38,368 +41,408 @@
...
+        StringBuilder sb = new StringBuilder();
+        Matcher mat = null;
+        // pstree pid get sub pids
+        if (OSUtils.isMacOS()) {
+            String pids = OSUtils.exeCmd("pstree -sp " + processId);
+            if (null != pids) {
+                mat = MACPATTERN.matcher(pids);
+            }
+        } else {
+            String pids = OSUtils.exeCmd("pstree -p " + processId);
```
Comment (lenboo) [inline dolphinscheduler-server/src/main/java/org/apache/dolphinscheduler/server/utils/ProcessUtils.java:399]: i think this "pstree " can be extracted to one variable,  then I can easily replace this variable.

**Review by lenboo — CHANGES_REQUESTED** [review]
other magic numbers exists.

**Review by lenboo — APPROVED** [review]
+1

**Review by lgcareer — APPROVED** [review]
+1

File: dolphinscheduler-server/src/main/java/org/apache/dolphinscheduler/server/utils/ProcessUtils.java [hunk 6]
```diff
@@ -38,368 +41,408 @@
...
+                FileUtils.writeStringToFile(new File(commandFile), sb.toString(), StandardCharsets.UTF_8);
+            }
+
+            String runCmd = String.format("%s %s", Constants.SH, commandFile);
+            if (StringUtils.isNotEmpty(tenantCode)) {
+                runCmd = "sudo -u " + tenantCode + " " + runCmd;
+            }
+
+            logger.info("kill cmd:{}", runCmd);
+            Runtime.getRuntime().exec(runCmd);
```
Comment (dailidong) [inline dolphinscheduler-server/src/main/java/org/apache/dolphinscheduler/server/utils/ProcessUtils.java:348]: please use OSUtils.exeCmd method instead，thx

**Review by dailidong — CHANGES_REQUESTED** [review]
don't use Runtime.getRuntime().exec directly, replace with OSUtils.exeCmd, thx

File: dolphinscheduler-server/src/main/java/org/apache/dolphinscheduler/server/utils/ProcessUtils.java (hunk 6 above)
Comment (Eights-Li) [reply dolphinscheduler-server/src/main/java/org/apache/dolphinscheduler/server/utils/ProcessUtils.java]: yes, fixed

**Comment by Eights-Li** [comment]
> don't use Runtime.getRuntime().exec directly, replace with OSUtils.exeCmd, thx
please review this pr~

**Review by dailidong — APPROVED** [review]
+1
good job

**Pull request merged**